### PR TITLE
TandemFoil: 6L/256d + grad-clip=1.0 (depth + stability)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
Gradient clipping proved to be a breakthrough on AirfRANS (7.3% improvement, PR #2680), where 91-98% of batches were being clipped with grad norms of 10-22. TandemFoil likely has similar gradient instability. We test grad-clip=1.0 while also adding a 6th layer (5L→6L) to explore whether extra depth helps when gradients are stable.

## Instructions
Run the following command exactly:

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil \
  --optimizer lion \
  --lr 3e-4 \
  --cosine-t-max 10 \
  --grad-clip 1.0 \
  --no-use-ema \
  --model-slices 64 \
  --model-layers 6 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --enable-fourier \
  --enable-te-coord-frame \
  --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure \
  --residual-prediction \
  --enable-pressure-prior-addition \
  --epochs 999 \
  --wandb-name tandem-6L256d-gradclip
```

Changes vs baseline: `--grad-clip 1.0` added, `--model-layers 6` (was 5).

## Baseline
- TandemFoil val: 52.81 / test: 55.25
- Baseline W&B run: `l5kggnbg`
- PR: #2595, 67 epochs, 180-min wall-clock
- Reproduce: `cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion --lr 3e-4 --cosine-t-max 10 --no-use-ema --model-slices 64 --model-layers 5 --model-hidden-dim 256 --model-heads 4 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999`

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2711 because ADVISOR: Closing — 6L/256d+gc at lr=3e-4 superseded by lr=2e-4..

**Comment by tcapelle** [comment]
STUDENT kakashi:

## Results

### Key Metrics

| Metric | Baseline (5L/256d, PR #2595) | This run (6L/256d + grad-clip=1.0) | Change |
|--------|------|------|--------|
| **val_primary/surface_pressure_mae** | 52.81 | **40.78** | **-22.8%** |
| **test_primary/surface_pressure_mae** | 55.25 | **40.44** | **-26.8%** |

### Per-Split Test Surface Pressure MAE

| Split | Baseline | This run | Change |
|-------|----------|----------|--------|
| single_in_dist | 61.80 | **37.25** | **-39.7%** |
| geom_camber_rc | 60.20 | **42.71** | **-29.1%** |
| geom_camber_cruise | 48.99 | **44.98** | **-8.2%** |
| re_rand | 50.00 | **36.81** | **-26.4%** |

### Gradient Clipping Analysis

| Epoch | Batches Clipped | Clip Rate | Avg Grad Norm |
|-------|----------------|-----------|---------------|
| 1 | 749/750 | 99.9% | 6.50 |
| 10 | 368/750 | 49.1% | 1.22 |
| 30 | 69/750 | 9.2% | 0.73 |
| 57 | 26/750 | 3.5% | 0.54 |

### Run Details
- **Command**: `cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 python train.py --dataset tandemfoil --optimizer lion --lr 3e-4 --cosine-t-max 10 --grad-clip 1.0 --no-use-ema --model-slices 64 --model-layers 6 --model-hidden-dim 256 --model-heads 4 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --agent kakashi --wandb-name "kakashi/tandem-6L256d-gradclip" --wandb-group tandem-6L256d-gradclip`
- **W&B run**: `pw7psent`
- **Epochs**: 57 (180-min timeout, still improving at cutoff)
- **Best epoch**: 57 (val 40.78)

### What happened

**Strong positive result.** 6L/256d with grad-clip=1.0 achieves a massive improvement over the 5L/256d baseline: -22.8% val, -26.8% test. This is a new TandemFoil best by a large margin.

Key observations:
1. **Gradient clipping was extremely active early on**: 99.9% of batches clipped at epoch 1 (avg grad norm 6.50), confirming TandemFoil has significant gradient instability like AirfRANS.
2. **Clipping rate declined naturally**: By epoch 57, only 3.5% of batches were clipped (avg grad norm 0.54). The model learned to produce smaller gradients, suggesting clipping helped most during early critical training phases.
3. **Late-training improvements**: Unlike the 5L baseline which plateaued at 67 epochs, this run kept improving strongly — best val dropped from 47.25 at epoch 35 to 40.78 at epoch 57. Still improving at cutoff.
4. **All splits improved uniformly**, with single_in_dist showing the largest gain (-39.7%).
5. **Confound note**: This is a two-variable change (6L + grad-clip). The grad clipping statistics suggest clipping was critical for early training stability, which likely enabled the deeper model to train more effectively. However, separate ablations would be needed to isolate each contribution.

**Note**: `test_geom_camber_cruise/volume_mae` reported Infinity — this appears to be a data issue (possibly a NaN in volume predictions for this split), but does not affect the primary surface_pressure_mae metrics.

### Suggested follow-ups
1. **Longer training**: Run was still improving at 57 epochs — a longer budget or higher epoch count could push below 40.
2. **Ablation: 5L/256d + grad-clip=1.0**: Isolate the grad-clip contribution at current best depth.
3. **Ablation: 6L/256d without grad-clip**: Isolate the depth contribution.
4. **Even deeper: 7L or 8L with grad-clip**: If 6L benefits from clipping, deeper models may benefit even more.
5. **Investigate Infinity in volume_mae**: The Infinity value in test_geom_camber_cruise/volume_mae should be investigated — could indicate numerical instability in volume predictions for this split.
6. **Lower clip threshold**: Try grad-clip=0.5 — research suggests tighter clipping may further stabilize Lion optimizer training.
